vllm/model_executor/parallel_utils/fast_allreduce.py

```diff
@@ -0,0 +1,169 @@
+
+
+@contextmanager
+def capture(enable: bool):
+    if enable:
+        init_fast_ar()
+    try:
+        begin_capture()
+        yield
+    finally:
+        end_capture()
+        if enable and get_handle() is not None:
+            get_handle().register_graph_buffers()
+
+
+# query if the set of gpus are fully connected by nvlink (1 hop)
+def _is_full_nvlink(rank, world_size):
+    pynvml.nvmlInit()
+    handle = pynvml.nvmlDeviceGetHandleByIndex(rank)
+    for i in range(world_size):
+        if i != rank:
+            try:
+                link_state = pynvml.nvmlDeviceGetNvLinkState(handle, i)
+                if not link_state:
+                    return False
+            except pynvml.NVMLError as error:
+                logger.info(
+                    f"NVLink detection failed with message \"{str(error)}\". "
+                    "This is normal if your machine has no NVLink equipped")
+                return False
+    pynvml.nvmlShutdown()
+    return True
+
+
+def _can_p2p(rank, world_size):
+    pynvml.nvmlInit()
+    handle1 = pynvml.nvmlDeviceGetHandleByIndex(rank)
+    for i in range(world_size):
+        if i != rank:
+            handle2 = pynvml.nvmlDeviceGetHandleByIndex(rank)
+            try:
+                p2p_status = pynvml.nvmlDeviceGetP2PStatus(
+                    handle1, handle2, pynvml.NVML_P2P_CAPS_INDEX_READ)
+                if p2p_status != pynvml.NVML_P2P_STATUS_OK:
+                    logger.info(
+                        f"P2P is not supported between device {i} and {rank}. "
+                        "Fast allreduce will be disabled")
+                    return False
+            except pynvml.NVMLError as error:
+                logger.info(
+                    f"P2P detection failed with message \"{str(error)}\". "
+                    "Fast allreduce will be disabled")
+                return False
+    pynvml.nvmlShutdown()
+    return True
+
+
+class FastAllreduce:
+
+    # max_size: max supported allreduce size
+    def __init__(self, rank, world_size, max_size=8192 * 1024) -> None:
+        # buffers memory are owned by this Python class and passed to C++
+        self.meta = torch.zeros(fast_ar.meta_size() + max_size,
+                                dtype=torch.uint8,
+                                device="cuda")
+        self.rank_data = torch.empty(16 * 1024 * 1024,
+                                     dtype=torch.uint8,
+                                     device="cuda")
+        self.max_size = max_size
+        self.world_size = world_size
+        handles, offsets = self._get_ipc_meta(self.meta)
+        self.full_nvlink = _is_full_nvlink(rank, world_size)
+        self._ptr = fast_ar.init_fast_ar(self.meta, self.rank_data, handles,
+                                         offsets, rank, self.full_nvlink)
+        self.fast_cond = self.full_nvlink or world_size <= 2
+
+    def _get_ipc_meta(self, inp: torch.Tensor):
+        data = inp.storage()._share_cuda_()
```

Comment:
With pytorch 2.1.2, I'm getting the warning
```
/home/ray/default/vllm/vllm/model_executor/parallel_utils/fast_allreduce.py:121: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  data = inp.storage()._share_cuda_()
```
here :)

Reply:
Fixed